update the Troubleshooting Automations page (#31810)
home-assistant/home-assistant.github.io: source/_docs/automation/troubleshooting.markdown

The right side of the trace screen has tabs with more information:

- Step Details shows data and results of the step that is currently highlighted.
- Automation Config shows the full YAML configuration at the time the automation was run.
- Trace Timeline, shown in the screenshot above, lists the steps that were executed and their timing.
- Related logbook entries, as the name says a logbook for all the entries related to the specific trace.
- Blueprint Config will only be shown if the automation was created from a {% term blueprint %}.

The top bar shows the date and time the automation was triggered. Use the left and right arrows to view previous runs of the automation.

Automations created in YAML must have an [`id`](/docs/automation/yaml/#migrating-your-yaml-automations-to-automationsyaml) assigned in order for debugging traces to be stored.

#### Traces
### Trace configuration

The last 5 traces are recorded for all automations. It is possible to change this by adding the following code to your automation.

{% raw %}

```yaml
trace:
    stored_traces: 1
  stored_traces: 20
```

[template]: /docs/configuration/templating/

## Testing your automation

> It is generally a difficult task to test an automation, especially if it includes several triggers and some conditions.

Reviewer: Why remove all this information about testing with triggers? I see that there is a more generic note above that triggers can't be tested that way. But for example the information that the test won't work if the automation uses data from triggers is no longer there. Was that intentional?

Reviewer 2: I tried to rearrange to present the simplest ways of testing first, the info here ended up near the top under "running the whole automation".  I tried to phrase things more "here's how to test these things" instead of starting with "this is difficult and trigger is useless".

Author: I tried to rearrange to present the simplest ways of testing first, the info here ended up near the top under "running the whole automation". I tried to phrase things more "here's how to test these things" instead of starting with "this is difficult and trigger is useless".

Reviewer: Yes, I saw that. And I agree with you. I also think it is better to first give instructions on what and how you can test. And then mention what is tricky. When I see that things are deleted, the first thing I do is check if information gets lost. So I'm wondering if it makes sense to add that back somewhere?

Reviewer: haha, just realized you beat me to it and added a section. Thx!

description: "Tips on how to troubleshoot your automations."
---

Automations and {% term scripts %} can be debugged in a few different ways. You can [test run](#testing-your-automation) the full sequence of actions, or test each condition and action separately. [Traces](#traces) let you see details of every step after an automation is run. For complicated automations with {% term templates %}, see the section [testing templates](#testing-templates).

## Testing your automation

Many automations can be tested directly in the automation editor UI.

> ### Running the whole automation

Reviewer: ```suggestion
### Running the entire automation
```

> In the 3-dots menu in the automation list or automation editor UI, click the **Run** button. This will execute all of the {% term actions %}, skipping all {% term triggers %} and {% term conditions %}. This lets you test the full sequence of actions, as if the automation was triggered and all conditions were true. 

Reviewer: ```suggestion
In the three dots menu in the automation list or automation editor UI, select the **Run** button. This executes all of the {% term actions %}, while skipping all {% term triggers %} and {% term conditions %}. This lets you test the full sequence of actions, as if the automation was triggered and all conditions were true. 
```

> In the 3-dots menu in the automation list or automation editor UI, click the **Run** button. This will execute all of the {% term actions %}, skipping all {% term triggers %} and {% term conditions %}. This lets you test the full sequence of actions, as if the automation was triggered and all conditions were true. Note that any [trigger ID](/automation/trigger/#trigger-id) used in your triggers will not be active when you test this way. The Trigger ID or any data passed by in the `trigger` data in conditions or actions can't be tested directly this way.

Reviewer: ```suggestion
In the three dots menu in the automation list or automation editor UI, select the **Run** button. This will execute all of the {% term actions %}, while skipping all {% term triggers %} and {% term conditions %}. This lets you test the full sequence of actions, as if the automation was triggered and all conditions were true. Note that any [trigger ID](/automation/trigger/#trigger-id) used in your triggers will not be active when you test this way. The Trigger ID or any data passed by in the `trigger` data in conditions or actions can't be tested directly this way.
```

Note that any [trigger ID](/automation/trigger/#trigger-id) used in your triggers will not be active when you test this way. The Trigger ID or any data passed by in the `trigger` data in conditions or actions can't be tested directly this way.

> You can also trigger an automation manually. This can test the conditions as if the automation was triggered by an event. Navigate to **{% my developer_services title="Developer tools -> Services" %}**. In the service selection drop-down, select "Automation: Trigger", then "Choose entity" to select the automation you are testing. Toggle whether to skip the conditions, then **CALL SERVICE**. If needed, additional `trigger` or other data can be added in the YAML view for testing.

Reviewer: ```suggestion
You can also trigger an automation manually. This can test the conditions as if the automation was triggered by an event. Navigate to {% my developer_services title="**Developer tools** > **Services**" %}. In the service selection drop-down, select **Automation: Trigger**, then **Choose entity** to select the automation you are testing. Toggle whether to skip the conditions, then **Call service**. If needed, additional `trigger` or other data can be added in the YAML view for testing.
```
Some tiny style tweaks:
- We use bold to markup UI strings.
- While we normally write the UI strings exactly as they are, we don't do the all caps. These could be a function that might be changed in a UI style change.

> You can also trigger an automation manually. This can test the conditions as if the automation was triggered by an event. Navigate to **{% my developer_services title="Developer tools -> Services" %}**. In the service selection drop-down, select "Automation: Trigger", then "Choose entity" to select the automation you are testing. Toggle whether to skip the conditions, then **CALL SERVICE**. If needed, additional `trigger` or other data can be added in the YAML view for testing. The [trigger](docs/automation/trigger/) page has more information about data within the trigger.

Reviewer: ```suggestion
You can also trigger an automation manually. This can test the conditions as if the automation was triggered by an event. Navigate to {% my developer_services title="**Developer tools** > **Services**" %}. In the service selection drop-down, select **Automation: Trigger**, then **Choose entity** to select the automation you are testing. Toggle whether to skip the conditions, then **Call service**. If needed, additional `trigger` or other data can be added in the YAML view for testing. The [trigger](docs/automation/trigger/) page has more information about data within the trigger.
```

### Running individual actions or conditions

> In the automation editor UI, each {% term condition %} and {% term action %} can be tested individually. Click the 3-dots menu, then the **Test** button.

Reviewer: ```suggestion
In the automation editor UI, each {% term condition %} and {% term action %} can be tested individually. Select the three dots menu, then the **Test** button.
```
I know *click* was used elsewhere in this doc. But in general, we use select because [click is mouse-oriented](https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/c/click).

- Testing a condition will highlight it to show whether the condition passed at the moment it was tested. If all conditions pass, then the automation will run when triggered. Testing building blocks like an **and** condition will report whether the whole block registers as true or false, or you can test individual conditions within the building block.

- Testing an action block will run that block immediately.

Note that complex automations that depend on previous blocks, such as trigger IDs, variables in templates, or service calls that return data to use in subsequent blocks, cannot be tested this way.

> If you are writing automations in YAML, it is also useful to go to **{% my server_controls title="Developer Tools -> YAML" %}** and click on **Check Configuration** button in Configuration validation section to make sure there are no syntax errors before restarting Home Assistant. In order for **Check configuration** to be visible, you must enable **Advanced Mode** on {% my profile title="your user profile" %}.

Reviewer: ```suggestion
If you are writing automations in YAML, it is also useful to go to {% my server_controls title="**Developer tools** > **YAML**" %}** and in the Configuration validation section, select the **Check configuration** button. This is to make sure there are no syntax errors before restarting Home Assistant. In order for **Check configuration** to be visible, you must enable **Advanced Mode** on {% my profile title="your user profile" %}.
```

## Traces

When an {% term automation %} is run, all steps are recorded and a trace is made. From the UI, open **Settings**, which is located in the sidebar, then select **Automations & Scenes** to go to the automation editor or click this button directly: {% my automations badge %}

> From the automation editor UI, or in the automations list in the 3-dots menu, click **Traces**. Alternatively, click on an automation entry shown in the Logbook.

Reviewer: ```suggestion
From the automation editor UI, or in the automations list in the three dots menu, select **Traces**. Alternatively, select an automation entry shown in the Logbook.
```

![Automation tracing example](/images/integrations/automation/automation-tracing.png)

The above screenshot shows a previous run of an automation. The automation is displayed using an interactive graph, highlighting which path the automation took. Each node in the graph can be clicked to view the details on what happened with the automation during that specific step. It traces the complete run of an automation.

The right side of the trace screen has tabs with more information:

> - Step Details shows data and results of the step that is currently highlighted.

Reviewer: ```suggestion
- **Step Details** shows data and results of the step that is currently highlighted.
```

> - Automation Config shows the full YAML configuration at the time the automation was run.

Reviewer: ```suggestion
- **Automation Config** shows the full YAML configuration at the time the automation was run.
```

> - Trace Timeline, shown in the screenshot above, lists the steps that were executed and their timing.

Reviewer: ```suggestion
- **Trace Timeline**, shown in the screenshot above, lists the steps that were executed and their timing.
```

> - Related logbook entries, as the name says a logbook for all the entries related to the specific trace.

Reviewer: ```suggestion
- **Related logbook entries**, shows a logbook for all the entries related to the specific trace.
```

> - Blueprint Config will only be shown if the automation was created from a {% term blueprint %}.

Reviewer: ```suggestion
- **Blueprint Config** will only be shown if the automation was created from a {% term blueprint %}.
```

The top bar shows the date and time the automation was triggered. Use the left and right arrows to view previous runs of the automation.

Automations created in YAML must have an [`id`](/docs/automation/yaml/#migrating-your-yaml-automations-to-automationsyaml) assigned in order for debugging traces to be stored.

### Trace configuration

The last 5 traces are recorded for all automations. It is possible to change this by adding the following code to your automation.

{% raw %}

```yaml
trace:
  stored_traces: 20
```

{% endraw %}

## Testing templates

If your automation uses [templates](/docs/configuration/templating/) in any part, you can do the following to make sure it works as expected:

> 1. Go to **{% my developer_template title="Developer tools -> Template" %}** tab.

Reviewer: ```suggestion
1. Go to {% my developer_template title="**Developer tools** > **Template**" %} tab.
```
